File: utils/matmul.py
import numpy as np 

def convertor(val , return_val = None):

    if isinstance(val , list) or isinstance(val , np.ndarray):
    
        if return_val : 
            if isinstance(val , list): val = np.array(val)
        else : 
            if isinstance(val , np.ndarray) : val = val.tolist()
    
    else : raise ValueError(f'Cannot process inputs of type {type(val)}, pass a numpy.ndarray or python list only')

    return val

def matmul(f_row , f_col , s_row , s_col):

    if len(f_col) != len(s_row) : raise ValueError(f'Cannot multiply array with dimensions {len(f_row)}x{len(f_col)} , {len(s_row)}x{len(s_col)}')

    f_row = convertor(f_row , return_val = True)
    f_col = convertor(f_col , return_val = True)
    s_row = convertor(s_row , return_val = True)
    s_col = convertor(s_col , return_val = True)

    val = sum(f_col * s_row / 2)
    col = s_col * val

    return col

File: utils/matmul.cpp
#include <iostream>
#include <vector>
using namespace std;

vector<double> matmul(vector<vector<double>>& arr, vector<vector<double>>& arr_) {
    vector<double> f_row = arr[0];
    vector<double> f_col = arr[1];

    vector<double> s_row = arr_[0];
    vector<double> s_col = arr_[1];

    double val = 0.0;
    for (size_t index = 0; index < f_col.size(); ++index) {
        val += (f_col[index] * s_row[index]) / 2.0;
    }

    vector<double> row;
    for (size_t index = 0; index < s_row.size(); ++index) {
        row.push_back(s_row[index] * val);
    }

    vector<double> col;
    for (size_t index = 0; index < s_col.size(); ++index) {
        col.push_back(s_col[index] * 2.0);
    }

    row.insert(row.end(), col.begin(), col.end());
    return row;
}

int main() {
    vector<vector<double>> arr = {{1.0, 2.0}, {3.0, 4.0}};
    vector<vector<double>> arr_ = {{5.0, 6.0}, {7.0, 8.0}};

    vector<double> result = matmul(arr, arr_);

    cout << "Result:";
    for (double value : result) {
        cout << " " << value;
    }
    cout << endl;

    return 0;
}
